### PR TITLE
add SetDecorations action

### LINKED ISSUE #1728: Cleaner approach to decorations
I believe the current approach to decorations is a bit messy and hard to extend. I tried to propose extensions in #1724 and #1725 with mixed success. If we want to avoid breaking changes in the future, I think it is better to think hard about the design first before merging any changes.

I am probably overthinking this. But now that the overthinking has already happened, I might as well share it.

## Why is this complicated?

We cannot directly port concepts from openbox because Wayland distinguishes between client side decorations (CSD) and server side decorations (SSD). 

## The current situation

As of 

- There are 3 modes of decoration: 'none', 'borders', and 'full'.
- The `ToggleDecoration` action is implemented, `Decorate` and `Undecorate` are not.
- CSD negotiation is implemented for both xdg-shell and xwayland.
  - The results can be overwritten using `<windowRule serverDecoration="">`
  - A default can be provided using `<core><decoration>`
  - CSD windows get initialized with 'none' decoration, SSD windows get initialized with 'full' decoration
- `<theme><keepBorder>` is implement, but behaves differently from openbox (and also from its documentation) (see #876)
  - If it is disabled, `ToggleDecorations` toggles between 'none' and 'full'
  - If it is enabled, `Toggledecorations` cycles between 'border', 'none', and 'full'
  - It has no effect on the initial decoration of CSD windows (as claimed by the documentation)
- The implementation doesn't use the 3 modes. Instead it uses the following:
  - `view->ssd_enabled` and `view->ssd_titlebar_hidden` to track whether ssd and titlebar are enabled
  - `view->ssd` and `view->ssd->titlebar.tree->node->enabled` to track whether ssd and titlebar actually exist. For example, full screen windows do not have decoration, even if it is enabled.
  - `has_ssd(view)` stores the result of the CSD/SSD negotiation

## What do we want?

As a start for discussion:

- We want to keep backwards compatibility as much as possible. Among other things, this means that we have to stick with the the 3-state `ToggleDecorations` and `keepBorder`.
- For CSD windows, undecorated means 'none'
- For SSD windows, the meaning of undecorated depends on `keepBorder`
  - if `keepBorder=yes`, undecorated means 'borders'
  - if `keepBorder=yes`, undecorated means 'none'
- Decorated always means 'full'
  - I am not 100% sure about this. Another approach would be to allow users to disable titlebars globally. In that case, decorated would mean 'border' and undecorated would mean 'none'. On the other hand, this could also be achieved by using something like `<windowRule serverDecoration="server"> <action name="Undecorate"/>` (rule) or `ToggleDecorations` (interactive).
  - You could also argue that for CSD windows, decorated should mean 'none' because they have their own decoration.
- Users should be able to switch to any of the 3 decoration modes manually, but not necessarily in a single step (this is what `ToggleDecorations` provides). 

## Settings / Rules / Actions

Actions are more flexible than rules (because they can be used both in `<windowRole>` and `<keybind>`/`<mousebind>`) and rules are more flexible than settings (because different rules can apply to different windows). So if we add anything, I think we should prefer actions over rules and settings.

## Use cases

Which of these do we want to support?

### disable CSD

```xml
<windowRule identifier="">
  <windowRule serverDecoration="server">
</windowRule>
```

### manually toggle decorations

```xml
<keybind key="">
  <action name="ToggleDecorations"/>
</keybind>
```

### manually toggle between 'full' and 'none'

Only possible if `keepBorder` is disabled globally.

### manually toggle between 'full' and 'border'

Not currently possible

### manually toggle between 'border' and 'none'

Not currently possible

### manually switch to 'full'

Not currently possible

### manually switch to 'border'

Not currently possible

### manually switch to 'none'

Not currently possible

### toggle between 'decorated' and 'undecorated' (whatever that means based on CSD/SSD negotation and `keepBorder`)

Not currently possible

### manually switch to 'decorated' (whatever that means based on CSD/SSD negotation and `keepBorder`)

Not currently possible

### manually switch to 'undecorated' (whatever that means based on CSD/SSD negotation and `keepBorder`)

Not currently possible

### disable titlebars

Not currently possible. It could look something like this:

```xml
<windowRule identifier="">
  <action name="Undecorate"/>
</windowRule>
```

### hide titlebars on maximize

Not currently possible (except that borders are removed automatically). It could look something like this:

```xml
<keybind key="W-Up">
  <action name="Undecorate"/>
  <action name="Maximize" direction="both"/>
</keybind>
<keybind key="W-Down">
  <action name="Decorate"/>
  <action name="Unmaximize"/>
</keybind>
<keybind key="W-Left">
  <action name="Undecorate"/>
  <action name="SnapToEdge" direction="left"/>
</keybind>
<keybind key="W-Right">
  <action name="Undecorate"/>
  <action name="SnapToEdge" direction="right"/>
</keybind>
<mousebind button="W-Left" action="Drag">
  <action name="Undecorate"/>
  <action name="Move"/>
</mousebind>
```

## Proposals

- Initially, `ToggleDecorations` toggled between 'none' and 'full'. This does not cover any usecases involving 'border', but it is straight forward and `Decorate`/`Undecorate` could easily be added.
- The current version was implemented in #876. It adds basic support for 'border', but does not cover all use cases. it is also no longer clear what `Decorate`/`Undecorate` should do.
- The initial version of #876 did not have the 3-state `ToggleDecorations`. Instead, it would toggle between 'full' and 'none' if `keepBorder=no` and between 'full' and 'border' otherwise. This was changed because it would add non-removable borders to CSD windows.
- In #1724 I proposed to mirror the structure of the code: `ToggleDecorations`/`Decorate`/`Undecorate` would manipulate `view->ssd_enabled`, and `ToggleTitlebar`/`ShowTitlebar`/`HideTitlebar` would manipulate `view->ssd_titlebar_hidden`. This would cover most use cases, but breaks backwards compatibility and was considered too low-level for end users.
- #1725 is based on a proposal by @ahesford.  The main idea was to add `keepBorder` attributes to some actions to overwrite the global setting. It covers many use cases without breaking backwards compatibility, but has some weird edge cases.
- `Decorate`/`Undecorate` could use rules based on `has_ssd(view)` and `keepBorder` to decide what is the right thing to do. That would be an opinionated approach that intentionally does not cover all use cases.

Some additional ideas that only cover specific use cases:

- There was also the proposal to add `<action name="SetServerSideDecorationMode" mode="none|border|all"/>`.
  - We could add value ranges, e.g. 'border-or-more' and 'border-or-less'. This would allow for example to remove titlebars from SSD windows without affecting CSD windows.
- The original use case why I got into all of this was disabling titlebars. That could be done with a configurable default value for `view->ssd_titlebar_hidden`. However, I would prefer to find a more generic solution.
- `<windowRule>` could gain an `decoration=server|client` attribute so that rules can be restricted to SSD or CSD windows.

### PR DESCRIPTION
**Pull request by xi**
as discussed in IRC. resolves #1728

We also discussed adding a "decorations" query for the if-action. I will work on that separately, but this already lays some ground work (e.g. `view_get_ssd_mode`).

### CODE REVIEW

**Comment by johanmalm** [comment]
I appreciate the work on this and noted the lengthy analysis. I haven’t been able to keep up with the story/thinking so if you want me to review anything, please could you point me to it.

If @ahesford or @Consolatis are already on the case I’m happy not to duplicate effort, but as I said if independent view is required (by me) please make it clear.

File: docs/labwc-actions.5.scd [hunk 1]
```diff
@@ -117,6 +117,16 @@ Actions are used in menus and keyboard/mouse bindings.
 	upper-left corner of the window associated with the action. Default is
 	yes.
 
+*<action name="SetDecorations" decorations="value" force_ssd="no" />*
```
Comment (Consolatis) [inline docs/labwc-actions.5.scd:120]: Suggest using `forceSSD="no"` to match the usual user-facing argument style.

File: docs/labwc-actions.5.scd [hunk 2]
```diff
@@ -117,6 +117,16 @@ Actions are used in menus and keyboard/mouse bindings.
...
+	Set decorations of focused window.
+
+	*decorations* [full|border|none] On border, only the titlebar will be
+	hidden, borders and resize area are kept. Default is full.
```
Comment (Consolatis) [inline docs/labwc-actions.5.scd:124]: This implies that the SSD is currently shown, suggest rewording it to something like

```
- *full* enables the whole server side decorations
- *border* only the border and invisible resize area are enabled
- *none* disables everything
```
It doesn't have to be a list.

File: include/view.h [hunk 3]
```diff
@@ -497,6 +505,7 @@ struct output *view_get_adjacent_output(struct view *view, enum view_edge edge,
 	bool wrap);
 enum view_axis view_axis_parse(const char *direction);
 enum view_edge view_edge_parse(const char *direction);
+enum ssd_mode ssd_mode_parse(const char *mode);
```
Comment (Consolatis) [inline include/view.h:508]: This function (and the enum) should likely live in `ssd.{h,c}`

File: include/ssd.h [hunk 4]
```diff
@@ -87,6 +87,7 @@ bool ssd_is_button(enum ssd_part_type type);
 bool ssd_part_contains(enum ssd_part_type whole, enum ssd_part_type candidate);
 
 /* TODO: clean up / update */
+bool has_ssd(struct view *view);
```
Comment (Consolatis) [inline include/ssd.h:90]: If this has to be public, it should be prefixed with `ssd_`.

Comment (ahesford) [reply include/ssd.h:90]: It's also misnamed, because the function indicates whether the view *wants* SSD, not whether it has them.

File: docs/labwc-actions.5.scd (hunk 1 above)
Comment (xi) [reply docs/labwc-actions.5.scd:120]: I had checked and found `output_name`. But you are right, camelcase that seems to be the exception.

Comment (Consolatis) [reply docs/labwc-actions.5.scd:120]: Whoops, that should likely be fixed then (e.g. updated in the docs, parsing `outputName` and logging an deprecation error when parsing output_name).

**Comment by johanmalm** [comment]
Thanks for working on this. Sorry it took a while to get into the merge window.
